File: src/samples/Sekiban.Dcb.Orleans.Decider.Wasm/SekibanDcbDecider.ClientApi/Endpoints/TestDataEndpoints.cs
using Dcb.EventSource.MeetingRoom.Queries;
using Dcb.EventSource.MeetingRoom.Room;
using Dcb.EventSource.MeetingRoom.User;
using Dcb.Interactions.Workflows.Reservation;
using Microsoft.AspNetCore.Mvc;
using Sekiban.Dcb;

namespace SekibanDcbDecider.ClientApi.Endpoints;

public static class TestDataEndpoints
{
    public static void MapTestDataEndpoints(this IEndpointRouteBuilder endpoints)
    {
        var group = endpoints.MapGroup("/test-data")
            .WithTags("Test Data")
            .RequireAuthorization();

        group.MapPost("/generate", GenerateTestDataAsync)
            .WithName("GenerateTestData");

        group.MapPost("/generate-rooms", GenerateRoomsAsync)
            .WithName("GenerateRooms");

        group.MapPost("/generate-reservations", GenerateReservationsAsync)
            .WithName("GenerateReservations");
    }

    private static async Task<IResult> GenerateTestDataAsync(
        [FromQuery] int? timeZoneOffsetMinutes,
        [FromServices] ISekibanExecutor executor,
        [FromServices] ILoggerFactory loggerFactory)
    {
        var logger = loggerFactory.CreateLogger("TestDataEndpoints");
        var result = new TestDataGenerationResult();

        // Generate sample user first
        var user = await GenerateUserInternalAsync(executor, logger);
        result.UserId = user.UserId;
        result.UserName = user.DisplayName;

        // Generate rooms
        var rooms = await GenerateRoomsInternalAsync(executor, logger);
        result.RoomsCreated = rooms.RoomIds.Count;
        result.RoomIds = rooms.RoomIds;
        result.SortableUniqueId = rooms.LastSortableUniqueId;
        await WaitForRoomsVisibleAsync(executor, rooms.RoomIds, rooms.LastSortableUniqueId, logger);

        // Generate reservations using the created rooms and user
        if (rooms.RoomIds.Count > 0)
        {
            var reservations = await GenerateReservationsInternalAsync(
                executor,
                rooms.RoomIds,
                user.UserId,
                user.DisplayName,
                timeZoneOffsetMinutes,
                logger);
            result.ReservationsCreated = reservations.ReservationIds.Count;
            result.ReservationIds = reservations.ReservationIds;
            result.Errors = reservations.Errors;
            result.SortableUniqueId = reservations.LastSortableUniqueId ?? result.SortableUniqueId;
            await WaitForReservationsVisibleAsync(
                executor,
                reservations.ReservationIds,
                reservations.LastSortableUniqueId,
                logger);
        }

        return Results.Ok(result);
    }

    private static async Task<IResult> GenerateRoomsAsync(
        [FromServices] ISekibanExecutor executor,
        [FromServices] ILoggerFactory loggerFactory)
    {
        var logger = loggerFactory.CreateLogger("TestDataEndpoints");
        var rooms = await GenerateRoomsInternalAsync(executor, logger);
        await WaitForRoomsVisibleAsync(executor, rooms.RoomIds, rooms.LastSortableUniqueId, logger);
        return Results.Ok(new
        {
            roomsCreated = rooms.RoomIds.Count,
            roomIds = rooms.RoomIds,
            sortableUniqueId = rooms.LastSortableUniqueId
        });
    }

    private static async Task<IResult> GenerateReservationsAsync(
        [FromQuery] Guid? roomId,
        [FromQuery] int? timeZoneOffsetMinutes,
        [FromServices] ISekibanExecutor executor,
        [FromServices] ILoggerFactory loggerFactory)
    {
        var logger = loggerFactory.CreateLogger("TestDataEndpoints");

        // Generate sample user first
        var user = await GenerateUserInternalAsync(executor, logger);

        // If no roomId provided, we need at least one room
        List<Guid> roomIds;
        if (roomId.HasValue)
        {
            roomIds = [roomId.Value];
        }
        else
        {
            // Try to create rooms first
            var rooms = await GenerateRoomsInternalAsync(executor, logger);
            await WaitForRoomsVisibleAsync(executor, rooms.RoomIds, rooms.LastSortableUniqueId, logger);
            roomIds = rooms.RoomIds;
            if (roomIds.Count == 0)
            {
                return Results.BadRequest("No rooms available for reservations");
            }
        }

        var reservations = await GenerateReservationsInternalAsync(
            executor,
            roomIds,
            user.UserId,
            user.DisplayName,
            timeZoneOffsetMinutes,
            logger);
        await WaitForReservationsVisibleAsync(
            executor,
            reservations.ReservationIds,
            reservations.LastSortableUniqueId,
            logger);
        return Results.Ok(new
        {
            reservationsCreated = reservations.ReservationIds.Count,
            reservationIds = reservations.ReservationIds,
            errors = reservations.Errors,
            sortableUniqueId = reservations.LastSortableUniqueId
        });
    }

    private static async Task<(Guid UserId, string DisplayName)> GenerateUserInternalAsync(
        ISekibanExecutor executor,
        ILogger? logger = null)
    {
        var userId = Guid.CreateVersion7();
        var displayName = "Sample User";
        var email = $"sample.user.{userId.ToString()[..8]}@example.com";

        try
        {
            logger?.LogInformation("Registering sample user: {Email} ({UserId})", email, userId);
            await executor.ExecuteAsync(new RegisterUser
            {
                UserId = userId,
                DisplayName = displayName,
                Email = email,
                Department = "Engineering"
            });
            logger?.LogInformation("Successfully registered sample user: {Email}", email);
        }
        catch (Exception ex)
        {
            logger?.LogWarning(ex, "Failed to register sample user (may already exist): {Email}", email);
        }

        return (userId, displayName);
    }

    private static async Task<GeneratedRoomsResult> GenerateRoomsInternalAsync(
        ISekibanExecutor executor,
        ILogger? logger = null)
    {
        var roomIds = new List<Guid>();
        string? lastSortableUniqueId = null;

        var roomDefinitions = new[]
        {
            new { Name = "Conference Room A", Capacity = 20, Location = "Building 1, Floor 2", Equipment = new List<string> { "Projector", "Whiteboard", "Video Conference" }, RequiresApproval = false },
            new { Name = "Meeting Room B", Capacity = 8, Location = "Building 1, Floor 3", Equipment = new List<string> { "TV Screen", "Whiteboard" }, RequiresApproval = false },
            new { Name = "Executive Boardroom", Capacity = 16, Location = "Building 2, Floor 5", Equipment = new List<string> { "Projector", "Video Conference", "Sound System", "Recording" }, RequiresApproval = true },
            new { Name = "Huddle Space 1", Capacity = 4, Location = "Building 1, Floor 1", Equipment = new List<string> { "TV Screen" }, RequiresApproval = false },
            new { Name = "Training Room", Capacity = 30, Location = "Building 3, Floor 1", Equipment = new List<string> { "Projector", "Multiple Screens", "Recording", "Microphones" }, RequiresApproval = true },
            new { Name = "Small Meeting Room C", Capacity = 6, Location = "Building 1, Floor 2", Equipment = new List<string> { "Whiteboard" }, RequiresApproval = false },
        };

        foreach (var roomDef in roomDefinitions)
        {
            try
            {
                var roomId = Guid.CreateVersion7();
                var command = new CreateRoom
                {
                    RoomId = roomId,
                    Name = roomDef.Name,
                    Capacity = roomDef.Capacity,
                    Location = roomDef.Location,
                    Equipment = roomDef.Equipment,
                    RequiresApproval = roomDef.RequiresApproval
                };

                logger?.LogDebug("Creating room: {RoomName} ({RoomId})", roomDef.Name, roomId);
                var result = await executor.ExecuteAsync(command);
                logger?.LogInformation("Created room: {RoomName} ({RoomId})", roomDef.Name, roomId);
                roomIds.Add(roomId);
                lastSortableUniqueId = result.SortableUniqueId ?? lastSortableUniqueId;
            }
            catch (Exception ex)
            {
                logger?.LogError(ex, "Failed to create room: {RoomName}", roomDef.Name);
            }
        }

        return new GeneratedRoomsResult(roomIds, lastSortableUniqueId);
    }

    private static async Task<GeneratedReservationsResult> GenerateReservationsInternalAsync(
        ISekibanExecutor executor,
        List<Guid> roomIds,
        Guid organizerId,
        string organizerName,
        int? timeZoneOffsetMinutes,
        ILogger? logger = null)
    {
        var reservationIds = new List<Guid>();
        var errors = new List<string>();
        string? lastSortableUniqueId = null;

        var (baseDateLocal, offset) = ResolveLocalBaseDate(timeZoneOffsetMinutes);
        var baseDate = baseDateLocal.AddDays(1);

        var reservationDefinitions = new[]
        {
            new { RoomIndex = 0, DaysOffset = 0, StartHour = 9, EndHour = 10, Purpose = "Team Standup", Confirm = true },
            new { RoomIndex = 0, DaysOffset = 0, StartHour = 14, EndHour = 16, Purpose = "Sprint Planning", Confirm = true },
            new { RoomIndex = 1, DaysOffset = 1, StartHour = 10, EndHour = 11, Purpose = "1:1 Meeting", Confirm = false },
            new { RoomIndex = 2, DaysOffset = 1, StartHour = 13, EndHour = 15, Purpose = "Board Meeting", Confirm = true },
            new { RoomIndex = 4, DaysOffset = 3, StartHour = 9, EndHour = 17, Purpose = "All-hands Training", Confirm = true },
        };

        var workflow = new QuickReservationWorkflow(executor);

        foreach (var resDef in reservationDefinitions)
        {
            try
            {
                // Skip if room index exceeds available rooms
                if (resDef.RoomIndex >= roomIds.Count)
                {
                    logger?.LogWarning("Skipping reservation '{Purpose}': room index {RoomIndex} exceeds available rooms ({RoomCount})",
                        resDef.Purpose, resDef.RoomIndex, roomIds.Count);
                    continue;
                }

                var roomId = roomIds[resDef.RoomIndex];
                var startTime = new DateTimeOffset(
                    baseDate.AddDays(resDef.DaysOffset).AddHours(resDef.StartHour),
                    offset).UtcDateTime;
                var endTime = new DateTimeOffset(
                    baseDate.AddDays(resDef.DaysOffset).AddHours(resDef.EndHour),
                    offset).UtcDateTime;

                logger?.LogDebug("Creating reservation '{Purpose}' for room {RoomId} from {StartTime} to {EndTime}",
                    resDef.Purpose, roomId, startTime, endTime);

                var result = await workflow.ExecuteAsync(
                    roomId,
                    organizerId,
                    organizerName,
                    startTime,
                    endTime,
                    resDef.Purpose,
                    null);

                logger?.LogInformation("Created reservation '{Purpose}' with ID {ReservationId}", resDef.Purpose, result.ReservationId);
                reservationIds.Add(result.ReservationId);
                lastSortableUniqueId = result.SortableUniqueId;
            }
            catch (Exception ex)
            {
                var errorMsg = $"Failed to create reservation '{resDef.Purpose}': {ex.Message}";
                errors.Add(errorMsg);
                logger?.LogError(ex, "Failed to create reservation '{Purpose}'", resDef.Purpose);
            }
        }

        if (errors.Count > 0)
        {
            logger?.LogWarning("Reservation generation completed with {ErrorCount} errors: {Errors}",
                errors.Count, string.Join("; ", errors));
        }

        return new GeneratedReservationsResult(reservationIds, errors, lastSortableUniqueId);
    }

    private static async Task WaitForRoomsVisibleAsync(
        ISekibanExecutor executor,
        IReadOnlyCollection<Guid> roomIds,
        string? waitForSortableUniqueId,
        ILogger? logger = null)
    {
        logger?.LogDebug(
            "Skipping room visibility wait for sample test data generation. RoomCount={RoomCount}, LastSortableUniqueId={WaitForSortableUniqueId}",
            roomIds.Count,
            waitForSortableUniqueId);
        await Task.CompletedTask;
    }

    private static async Task WaitForReservationsVisibleAsync(
        ISekibanExecutor executor,
        IReadOnlyCollection<Guid> reservationIds,
        string? waitForSortableUniqueId,
        ILogger? logger = null)
    {
        logger?.LogDebug(
            "Skipping reservation visibility wait for sample test data generation. ReservationCount={ReservationCount}, LastSortableUniqueId={WaitForSortableUniqueId}",
            reservationIds.Count,
            waitForSortableUniqueId);
        await Task.CompletedTask;
    }

    private static (DateTime baseDateLocal, TimeSpan offset) ResolveLocalBaseDate(int? timeZoneOffsetMinutes)
    {
        var offset = timeZoneOffsetMinutes.HasValue
            ? TimeSpan.FromMinutes(-timeZoneOffsetMinutes.Value)
            : DateTimeOffset.Now.Offset;
        var nowLocal = DateTimeOffset.UtcNow.ToOffset(offset);
        return (nowLocal.Date, offset);
    }
}

public record TestDataGenerationResult
{
    public Guid? UserId { get; set; }
    public string? UserName { get; set; }
    public int RoomsCreated { get; set; }
    public List<Guid> RoomIds { get; set; } = [];
    public int ReservationsCreated { get; set; }
    public List<Guid> ReservationIds { get; set; } = [];
    public List<string> Errors { get; set; } = [];
    public string? SortableUniqueId { get; set; }
}

public record GeneratedRoomsResult(List<Guid> RoomIds, string? LastSortableUniqueId);

public record GeneratedReservationsResult(List<Guid> ReservationIds, List<string> Errors, string? LastSortableUniqueId);

File: src/samples/Sekiban.Dcb.Orleans.Decider.Wasm.Swift/SekibanDcbDecider.Swift.ClientApi/Sources/SekibanDcbDeciderSwiftClientApi/TestDataEndpoints.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif
import HTTPTypes
import Hummingbird
import Logging
import NIOCore
import SekibanDcbDeciderSwiftClientApiCore

// /api/test-data/* — creates a demo corpus of rooms + reservations in one call, matching
// the shape the WebNext frontend's `testData.generate` tRPC procedure expects. The native
// template's equivalent lives in SekibanDcbDecider.ApiService/Endpoints/TestDataEndpoints.cs;
// we reuse the existing CreateRoom / CreateQuickReservation builders so event shapes stay
// identical to every other write path in this sample.

func registerTestDataRoutes(
    _ router: Router<BasicRequestContext>,
    wasmServerUrl: String,
    logger: Logger
) {
    let forwarder = CommitRequestForwarder(wasmServerUrl: wasmServerUrl)
    let reader = TagLatestSortableReader(wasmServerUrl: wasmServerUrl)

    router.post("/api/test-data/generate") { request, _ in
        let tzMinutes = queryInt(request, key: "timeZoneOffsetMinutes")
        return try await generateTestDataResponse(
            forwarder: forwarder, reader: reader, logger: logger,
            wasmServerUrl: wasmServerUrl,
            timeZoneOffsetMinutes: tzMinutes,
            includeReservations: true,
            onlyRoomIds: nil)
    }

    router.post("/api/test-data/generate-rooms") { _, _ in
        try await generateTestDataResponse(
            forwarder: forwarder, reader: reader, logger: logger,
            wasmServerUrl: wasmServerUrl,
            timeZoneOffsetMinutes: nil,
            includeReservations: false,
            onlyRoomIds: nil)
    }

    router.post("/api/test-data/generate-reservations") { request, _ in
        let tzMinutes = queryInt(request, key: "timeZoneOffsetMinutes")
        let explicitRoomId = queryString(request, key: "roomId").flatMap(UUID.init(uuidString:))
        let forcedRooms: [UUID]? = explicitRoomId.map { [$0] }
        return try await generateTestDataResponse(
            forwarder: forwarder, reader: reader, logger: logger,
            wasmServerUrl: wasmServerUrl,
            timeZoneOffsetMinutes: tzMinutes,
            includeReservations: true,
            onlyRoomIds: forcedRooms)
    }
}

// ---------------------------------------------------------------------------
// Core helper
// ---------------------------------------------------------------------------

private struct TestDataResult: Encodable {
    var roomsCreated: Int
    var roomIds: [String]
    var reservationsCreated: Int
    var reservationIds: [String]
    var errors: [String]
}

private let roomDefinitions: [(name: String, capacity: Int32, location: String, equipment: [String], requiresApproval: Bool)] = [
    ("Conference Room A", 20, "Building 1, Floor 2", ["Projector", "Whiteboard", "Video Conference"], false),
    ("Meeting Room B", 8, "Building 1, Floor 3", ["TV Screen", "Whiteboard"], false),
    ("Executive Boardroom", 16, "Building 2, Floor 5", ["Projector", "Video Conference", "Sound System", "Recording"], true),
    ("Huddle Space 1", 4, "Building 1, Floor 1", ["TV Screen"], false),
    ("Training Room", 30, "Building 3, Floor 1", ["Projector", "Multiple Screens", "Recording", "Microphones"], true),
    ("Small Meeting Room C", 6, "Building 1, Floor 2", ["Whiteboard"], false),
]

private let reservationDefinitions: [(roomIndex: Int, daysOffset: Int, startHour: Int, endHour: Int, purpose: String)] = [
    (0, 0, 9, 10, "Team Standup"),
    (0, 0, 14, 16, "Sprint Planning"),
    (1, 1, 10, 11, "1:1 Meeting"),
    (2, 1, 13, 15, "Board Meeting"),
    (4, 3, 9, 17, "All-hands Training"),
]

private func generateTestDataResponse(
    forwarder: CommitRequestForwarder,
    reader: TagLatestSortableReader,
    logger: Logger,
    wasmServerUrl: String,
    timeZoneOffsetMinutes: Int?,
    includeReservations: Bool,
    onlyRoomIds: [UUID]?
) async throws -> Response {
    var result = TestDataResult(
        roomsCreated: 0, roomIds: [], reservationsCreated: 0,
        reservationIds: [], errors: [])
    var lastSortableUniqueId: String?

    let roomIds: [UUID]
    if let forced = onlyRoomIds {
        roomIds = forced
    } else {
        var created: [UUID] = []
        for def in roomDefinitions {
            do {
                let build = try buildCreateRoomCommit(
                    request: CreateRoomRequest(
                        roomId: UUID(),
                        name: def.name,
                        capacity: def.capacity,
                        location: def.location,
                        equipment: def.equipment,
                        requiresApproval: def.requiresApproval))
                let commitResult = try await forwarder.forward(build.request)
                if commitResult.isSuccess {
                    created.append(build.roomId)
                    if let suid = commitResult.sortableUniqueId { lastSortableUniqueId = suid }
                } else {
                    let body = String(decoding: commitResult.body, as: UTF8.self)
                    result.errors.append("room '\(def.name)' status=\(commitResult.statusCode): \(body)")
                }
            } catch {
                result.errors.append("room '\(def.name)' failed: \(error)")
            }
        }
        roomIds = created
    }
    result.roomsCreated = roomIds.count
    result.roomIds = roomIds.map { $0.uuidString.lowercased() }

    if includeReservations, !roomIds.isEmpty {
        let organizerId = UUID()
        let organizerName = "Sample User"
        let (baseDate, offsetSeconds) = resolveLocalBaseDate(timeZoneOffsetMinutes: timeZoneOffsetMinutes)
        let firstReservableDay = baseDate.addingTimeInterval(24 * 60 * 60)

        for def in reservationDefinitions {
            guard def.roomIndex < roomIds.count else {
                logger.warning("skipping reservation '\(def.purpose)': room index \(def.roomIndex) ≥ \(roomIds.count)")
                continue
            }
            let roomId = roomIds[def.roomIndex]
            let start = firstReservableDay
                .addingTimeInterval(Double(def.daysOffset) * 24 * 60 * 60)
                .addingTimeInterval(Double(def.startHour) * 60 * 60)
            let end = firstReservableDay
                .addingTimeInterval(Double(def.daysOffset) * 24 * 60 * 60)
                .addingTimeInterval(Double(def.endHour) * 60 * 60)
            // Convert from local wall-clock back to UTC by subtracting the offset.
            let startUtc = start.addingTimeInterval(-offsetSeconds)
            let endUtc = end.addingTimeInterval(-offsetSeconds)

            do {
                let build = try await buildCreateQuickReservationCommit(
                    request: CreateQuickReservationRequest(
                        reservationId: UUID(),
                        roomId: roomId,
                        organizerId: organizerId,
                        organizerName: organizerName,
                        startTime: SekibanTime.iso8601(from: startUtc),
                        endTime: SekibanTime.iso8601(from: endUtc),
                        attendeeCount: nil,
                        purpose: def.purpose,
                        selectedEquipment: nil),
                    reader: reader)
                let commitResult = try await forwarder.forward(build.request)
                if commitResult.isSuccess {
                    result.reservationsCreated += 1
                    result.reservationIds.append(build.reservationId.uuidString.lowercased())
                    if let suid = commitResult.sortableUniqueId { lastSortableUniqueId = suid }
                } else {
                    let body = String(decoding: commitResult.body, as: UTF8.self)
                    result.errors.append("reservation '\(def.purpose)' status=\(commitResult.statusCode): \(body)")
                }
            } catch {
                result.errors.append("reservation '\(def.purpose)' failed: \(error)")
            }
        }
    }

    // Block until the Room and Reservation projections have consumed the last commit.
    // Without this the Next.js UI's refetch races the projector and shows 0 rooms /
    // 0 reservations for a second or two after a successful "Generate Test Data" click.
    if let suid = lastSortableUniqueId {
        await waitForProjectionCatchUp(
            wasmServerUrl: wasmServerUrl,
            queryType: "GetRoomListQuery",
            sortableUniqueId: suid,
            logger: logger)
        if includeReservations {
            await waitForProjectionCatchUp(
                wasmServerUrl: wasmServerUrl,
                queryType: "GetReservationListQuery",
                sortableUniqueId: suid,
                logger: logger)
        }
    }

    let data = try JSONEncoder().encode(result)
    var buffer = ByteBuffer()
    buffer.writeBytes(data)
    return Response(
        status: .ok,
        headers: [.contentType: "application/json"],
        body: ResponseBody(byteBuffer: buffer))
}

// Returns (midnight-local-today-as-UTC-instant, secondsEastOfUtc) so callers can
// compose local hours and then convert back to UTC by subtracting the offset.
private func resolveLocalBaseDate(timeZoneOffsetMinutes: Int?) -> (Date, Double) {
    // JS `Date.getTimezoneOffset()` returns UTC-minus-local in minutes — so a UTC caller
    // sends 0, and a JST (UTC+9) caller sends -540. Invert the sign to get seconds east
    // of UTC, matching the C# template's convention.
    let offsetSeconds: Double
    if let m = timeZoneOffsetMinutes {
        offsetSeconds = Double(-m) * 60.0
    } else {
        offsetSeconds = Double(TimeZone.current.secondsFromGMT())
    }
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = TimeZone(secondsFromGMT: Int(offsetSeconds)) ?? .current
    let startOfDay = calendar.startOfDay(for: Date())
    return (startOfDay, offsetSeconds)
}

/// Issues a `/api/sekiban/serialized/list-query` with `waitForSortableUniqueId` set and
/// ignores the body. This is how we force the wasmserver's projection manager to apply
/// the given event before returning from /api/test-data/generate — without it, the
/// Next.js UI refetches before the projector catches up and shows an empty list.
/// Uses the same `URLSession.shared.data(for:)` path as `QueryForwarder.post` so
/// behavior is identical on macOS and Linux (`FoundationNetworking`).
private func waitForProjectionCatchUp(
    wasmServerUrl: String,
    queryType: String,
    sortableUniqueId: String,
    logger: Logger
) async {
    guard let url = URL(string: wasmServerUrl.trimmingSlash() + "/api/sekiban/serialized/list-query") else {
        return
    }
    var urlRequest = URLRequest(url: url)
    urlRequest.httpMethod = "POST"
    urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
    let payload: [String: Any] = [
        "queryType": queryType,
        "queryParamsJson": "{}",
        "waitForSortableUniqueId": sortableUniqueId,
    ]
    urlRequest.httpBody = (try? JSONSerialization.data(withJSONObject: payload))
        ?? Data(#"{"queryType":"\#(queryType)","queryParamsJson":"{}","waitForSortableUniqueId":"\#(sortableUniqueId)"}"#.utf8)
    do {
        // URLSession does not throw on non-2xx — inspect the response explicitly so a
        // 4xx/5xx from wasmserver doesn't silently masquerade as a successful catch-up.
        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        if let http = response as? HTTPURLResponse, http.statusCode >= 400 {
            let text = String(data: data, encoding: .utf8) ?? "<non-utf8>"
            logger.warning(
                "waitForProjectionCatchUp(\(queryType)) → \(http.statusCode) \(text)")
        }
    } catch {
        logger.warning("waitForProjectionCatchUp(\(queryType)) failed: \(error)")
    }
}

private func queryInt(_ request: Request, key: String) -> Int? {
    guard let raw = request.uri.queryParameters[Substring(key)] else { return nil }
    return Int(String(raw))
}

private func queryString(_ request: Request, key: String) -> String? {
    guard let raw = request.uri.queryParameters[Substring(key)] else { return nil }
    return String(raw)
}
